plugins/inputs/intel_powerstat/services/msr.go

```diff
@@ -0,0 +1,297 @@
+	c3StateResidencyLocation           = "0x3FC"
+	c6StateResidencyLocation           = "0x3FD"
+	c7StateResidencyLocation           = "0x3FE"
+	maximumFrequencyClockCountLocation = "0xE7"
+	actualFrequencyClockCountLocation  = "0xE8"
+	throttleTemperatureLocation        = "0x1A2"
+	temperatureLocation                = "0x19C"
+	// TimestampCounterLocation is an MSR offset.
+	TimestampCounterLocation = "0x10"
+)
+
+// MsrService is responsible for interactions with MSR.
+type MsrService interface {
+	OpenAndReadMsr(core string) error
+	RetrieveCPUFrequencyForCore(core string) (float64, int64, error)
+	GetCpuIDs() []string
+	GetThrottleTemperature(CPU string) uint64
+	GetTemperature(CPU string) uint64
+	GetTimestampDelta(CPU string) uint64
+	GetMperfDelta(CPU string) uint64
+	GetAperfDelta(CPU string) uint64
+	GetC6Delta(CPU string) uint64
+	GetC3Delta(CPU string) uint64
+	GetC7Delta(CPU string) uint64
+	GetReadDate(CPU string) int64
+	SetReadDate(CPU string, date int64)
+}
+
+// MsrServiceImpl is implementation of MsrService.
+type MsrServiceImpl struct {
+	cpuCoresData map[string]*data.MsrData
+	msrOffsets   map[string]int64
+	fs           FileService
+	log          telegraf.Logger
+}
+
+// GetCpuIDs returns array of logical CPUs.
+func (m *MsrServiceImpl) GetCpuIDs() []string {
+	IDs := make([]string, 0)
+
+	for cpuID := range m.cpuCoresData {
+		IDs = append(IDs, cpuID)
+	}
+
+	return IDs
+}
+
+// GetThrottleTemperature returns throttle temperature for CPU.
+func (m *MsrServiceImpl) GetThrottleTemperature(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].ThrottleTemp
+}
+
+// GetTemperature returns temperature for CPU.
+func (m *MsrServiceImpl) GetTemperature(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].Temp
+}
+
+// GetTimestampDelta returns timestamp delta for CPU.
+func (m *MsrServiceImpl) GetTimestampDelta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].TscDelta
+}
+
+// GetMperfDelta returns mperf delta for CPU.
+func (m *MsrServiceImpl) GetMperfDelta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].MperfDelta
+}
+
+// GetAperfDelta returns aperf delta for CPU.
+func (m *MsrServiceImpl) GetAperfDelta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].AperfDelta
+}
+
+// GetC6Delta returns C6 delta for CPU.
+func (m *MsrServiceImpl) GetC6Delta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].C6Delta
+}
+
+// GetC3Delta returns C3 delta for CPU.
+func (m *MsrServiceImpl) GetC3Delta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].C3Delta
+}
+
+// GetC7Delta returns C7 delta for CPU.
+func (m *MsrServiceImpl) GetC7Delta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].C7Delta
+}
+
+// GetReadDate returns last read date.
+func (m *MsrServiceImpl) GetReadDate(cpuID string) int64 {
+	return m.cpuCoresData[cpuID].ReadDate
+}
+
+// SetReadDate sets last read date.
+func (m *MsrServiceImpl) SetReadDate(cpuID string, date int64) {
+	m.cpuCoresData[cpuID].ReadDate = date
+}
```

Comment:
I don't think we need this. The fields are public anyway and we are the only internal user.

Reply:
You're right. Removed.